department-of-veterans-affairs/abd-vro
# VRO Console: monitor Camel messages

## What was the problem?


As a follow-on to #531, we'd like monitor claims as they are being processed along Camel routes for diagnosing problems particularly in prod.
The RabbitMQ Management UI is not sufficient.

Associated tickets or Slack threads:

- #562 

## How does this fix it?


Adds [Camel wiretaps](https://camel.apache.org/components/3.18.x/eips/wireTap-eip.html) at critical points in Camel routes to monitor messages and their content.

## How to test this PR
- `export COMPOSE_PROFILES=debug` -- the console container only starts if the docker-compose `debug` profile is enabled.
- Start VRO. Note the `vro-console-1` container; check out the logs: `docker logs vro-console-1`.
- Attach to the container: `docker attach vro-console-1`
- On the `groovy:000>` prompt, try the following:
```groovy
// Check out http://localhost:15672/#/queues for current queues and see how `console-*` queues are added as the following commands are run.

// Initially no routes in the CamelContext of the Console container
groovy:000> camel.routes
===> []

groovy:000> wireTap claim-submitted
// Now, submit a claim using Swagger
// Expect to see a log message

groovy:000> wireTap generate-pdf
// Now, generate a pdf using Swagger
// Expect to see a log message

groovy:000> camel.routes
===> [
  Route[rabbitmq://tap-generate-pdf?exchangeType=topic&queue=console-generate-pdf -> null], 
  Route[rabbitmq://tap-claim-submitted?exchangeType=topic&queue=console-claim-submitted -> null]
]
```
- Stopping the Console container will remove the `console-*` queues.
- Check that the `app` container logs has no errors.
